File: 08. Trees/16. Symmetric Tree/main.cpp
#include <bits/stdc++.h>
using namespace std;

class TreeNode {
public:
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

// DFS Approach
bool isMirror(TreeNode* t1, TreeNode* t2) {
    if (!t1 && !t2) return true;  // Both null -> symmetric
    if (!t1 || !t2) return false; // One null -> not symmetric
    return (t1->val == t2->val) &&
           isMirror(t1->left, t2->right) &&
           isMirror(t1->right, t2->left);
}

bool isSymmetricDFS(TreeNode* root) {
    if (!root) return true;
    return isMirror(root->left, root->right);
}

// BFS Approach
bool isSymmetricBFS(TreeNode* root) {
    if (!root) return true;

    queue<pair<TreeNode*, TreeNode*>> q;
    q.push({root->left, root->right});

    while (!q.empty()) {
        auto [t1, t2] = q.front();
        q.pop();

        if (!t1 && !t2) continue;
        if (!t1 || !t2 || t1->val != t2->val) return false;

        q.push({t1->left, t2->right});
        q.push({t1->right, t2->left});
    }
    return true;
}

// Main function
int main() {
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(2);
    root->left->left = new TreeNode(3);
    root->left->right = new TreeNode(4);
    root->right->left = new TreeNode(4);
    root->right->right = new TreeNode(3);

    cout << "DFS Approach: " << (isSymmetricDFS(root) ? "True" : "False") << endl;
    cout << "BFS Approach: " << (isSymmetricBFS(root) ? "True" : "False") << endl;

    return 0;
}

File: 08. Trees/16. Symmetric Tree/main.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left, right;
    TreeNode(int x) { val = x; left = right = null; }
}

public class main {

    // DFS Approach
    public static boolean isMirrorDFS(TreeNode t1, TreeNode t2) {
        if (t1 == null && t2 == null) return true;
        if (t1 == null || t2 == null) return false;
        return (t1.val == t2.val) && 
               isMirrorDFS(t1.left, t2.right) &&
               isMirrorDFS(t1.right, t2.left);
    }

    public static boolean isSymmetricDFS(TreeNode root) {
        if (root == null) return true;
        return isMirrorDFS(root.left, root.right);
    }

    // BFS Approach
    public static boolean isSymmetricBFS(TreeNode root) {
        if (root == null) return true;
        
        Queue<TreeNode[]> queue = new LinkedList<>();
        queue.add(new TreeNode[]{root.left, root.right});

        while (!queue.isEmpty()) {
            TreeNode[] pair = queue.poll();
            TreeNode t1 = pair[0], t2 = pair[1];

            if (t1 == null && t2 == null) continue;
            if (t1 == null || t2 == null || t1.val != t2.val) return false;

            queue.add(new TreeNode[]{t1.left, t2.right});
            queue.add(new TreeNode[]{t1.right, t2.left});
        }
        return true;
    }

    // Main Function
    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(2);
        root.left.left = new TreeNode(3);
        root.left.right = new TreeNode(4);
        root.right.left = new TreeNode(4);
        root.right.right = new TreeNode(3);

        System.out.println("DFS Approach: " + isSymmetricDFS(root));
        System.out.println("BFS Approach: " + isSymmetricBFS(root));
    }
}
